### PR TITLE
Fix gradient NaN issues in sigmoid_focal_loss for extreme logits

### LINKED ISSUE #1267: NaNs with optax.losses.sigmoid_focal_loss
I've started seeing model instability after using `optax.losses.sigmoid_focal_loss()`, with NaNs occurring in gradients several thousand steps into training.

This occurs when `alpha = None` and `gamma = 0.0`, which I understand should be equivalent to `optax.losses.sigmoid_binary_cross_entropy()`. When `gamma > 0`, the model trains correctly.

### PR DESCRIPTION
**Pull request by leochlon**
addresses numerical instability in gradient and Hessian computation when using sigmoid_focal_loss with large logit values and gamma < 2.  might fix #1267 

the issue occurs because (1 - p_t) approaches zero for extreme logits, causing instability in derivatives of (1 - p_t)^gamma when gamma < 2.

clamping (1 - p_t) to be at least machine epsilon, preventing division by zero and infinite derivatives while maintaining mathematical equivalence for normal cases is a fix

this enables stable use of focal loss with Newton's method, L-BFGS, and other second-order optimization algorithms across the full range of valid gamma values.

test plan:  

logits = jnp.array([100.0])
labels = jnp.array([1.0])

Jacobian test (γ=0.5)
grad = jax.grad(lambda x: sigmoid_focal_loss(x, labels, gamma=0.5)[0])(logits)[0]
print(f"Jacobian finite: {jnp.isfinite(grad)}")

Hessian test (γ=1.5)
hess = jax.hessian(lambda x: sigmoid_focal_loss(x, labels, gamma=1.5)[0])(logits)[0, 0]
print(f"Hessian finite: {jnp.isfinite(hess)}")

python -m pytest optax/losses/_classification_test.py::SigmoidFocalLossTest -v

passes

### CODE REVIEW

**Review by Vishal-sys-code — APPROVED** [review]
Excellent fix, unconditionally clamping `(1 – pₜ)` to ε ensures stable gradients/Hessians for all γ > 0. Only tweak: add a brief docstring note that “`(1–pₜ)` is clamped to ε to guarantee finite derivatives.” Otherwise, LGTM!

**Comment by leochlon** [comment]
@Vishal-sys-code  I appreciate you taking the time to review the PR  so quickly! Thank you for the great feedback, I've added a brief note to the docstring explaining the clamping for finite derivatives as you suggested. Tests are now passing ready to merge :)

**Review by Vishal-sys-code — APPROVED** [review]

**Comment by rdyro** [comment]
It seems you still have some files in the root directory in the history. You can also run the linter locally using either `ruff check .` or `pre-commit run -a`

**Comment by leochlon** [comment]
Thanks @rdyro, I'll note this for next time! :)

**Comment by leochlon** [comment]
@vroulet , please let me know what you think of the implementation and if it matches what you were thinking. 

Image

Image

**Comment by vroulet** [comment]
Ok @leochlon are you vibe coding?
How do you expect us maintainers (mostly volunteers) to interact with you?
(It's really an open question, not a rhetoric one.)

**Comment by vroulet** [comment]
To be clear, I like this contribution, and you understood pretty well the mathematics behind this I think.
But a library is supposed to provide clear, concise, clean code. You provided 500 lines of code.
Do you expect us to review all 500 lines? 

Again this is an open question, and your answer will truly be appreciated. You won't be the last one to vibe code and you may be the first one to actually care and produce something interesting for the library.

**Comment by leochlon** [comment]
@vroulet 

Thanks for the feedback - let me clarify my process and intentions:

On the 500 lines: A significant portion is the automated stability test, which I included as a reproducible proof of the numerical properties but didn't intend for the final merge. I wanted to demonstrate that the implementation handles the edge cases correctly, but I realise dumping it all in one PR makes review difficult.

I spent about 2 days working through the mathematics and implementation. I do use Copilot for debugging and code cleanup, but the core logic and mathematical understanding is mine. I can explain any part of the algorithm and respond to technical feedback.

**Comment by leochlon** [comment]
@vroulet 

What I'd actually propose for the library:

The core log-space implementation (~20-30 lines)

Basic correctness tests + a simple regression test for numerical stability

If you'd like, I can split this into a focused PR with just the core changes + basic tests, with the stability analysis as supporting documentation. 

Would you prefer the core implementation as a clean, minimal PR with the analysis referenced separately?

**Comment by vroulet** [comment]
@leochlon thank you for the answer! Again I appreciate this contribution, it clears up a problem that has been lingering for some time in a very clean way.

Yes, ideally just the core log-space implementation and associated tests would be the best.

About supporting documentation for future PRs. The best practice would be to add it comments to the PR as a reproducible code example. That way maintainers can easily run the example without having to carefully read each line. Careful reviews are really important for good maintenance and so all code that is in your PR should generally be carefully read by someone (like me for example) so it's best to keep it short for the sake of the maintainers.

Thank you again. Sorry if I was very direct but some previous users did not care as much as you and simply returned llm generated code without even communicating (you did and I sincerely appreciate it).

**Comment by leochlon** [comment]
@vroulet I really appreciate your dedication keeping this repo high quality, there's so much cool stuff to build with optax and I've been working on a bunch of other improvements, so hopefully we can work together more going forward.

Makes total sense about keeping PRs focused and reviewable. I'll restructure this as:

Clean core implementation + tests in the PR
Put the extensive validation stuff in comments as reproducible examples

I'm pretty obsessive about reproducibility and including comprehensive tests to back up why I change things, but now I know the right way to present them without creating review burden (again sorry about that haha).

Thanks for being welcoming and explaining the process! I'll clean this up to just the core log-space fix and move the stability analysis to comments. Looking forward to contributing more here.

File: optax/losses/_classification.py [hunk 1]
```diff
@@ -814,47 +815,88 @@ def sigmoid_focal_loss(
...
   References:
-    Lin et al. `Focal Loss for Dense Object Detection
+    Lin et al, `Focal Loss for Dense Object Detection
     <https://arxiv.org/abs/1708.02002>`_, 2017
+
+  .. versionchanged:: 0.2.5
+    Added numerical stability improvements using log-space computation.
+    Added support for continuous labels in `[0, 1]`.
   """
   alpha = -1 if alpha is None else alpha
```
Comment (vroulet) [inline optax/losses/_classification.py]: As you're on it, this piece of code is strange. 
At the end of the code there should simply be a "if alpha is not None, ....".
No need to convert alpha to a negative number I think.

File: optax/losses/_classification.py [hunk 2]
```diff
@@ -814,47 +815,88 @@ def sigmoid_focal_loss(
...
-  loss = jnp.where(alpha >= 0, weighted, loss)
+  log_one_minus_p_t = jax.scipy.special.logsumexp(
+      jnp.stack([log_p, log_q], axis=-1),
+      axis=-1,
+      b=jnp.stack([1 - labels, labels], axis=-1)
+  )
+
+  # Gradient stability clamping
+  eps = jnp.finfo(logits.dtype).eps
+  log_one_minus_p_t_safe = jnp.maximum(log_one_minus_p_t, jnp.log(eps))
```
Comment (vroulet) [inline optax/losses/_classification.py]: Do you still need the clamping or is it really for unreasonable cases (like logits up 10**6)?

File: optax/losses/_classification_test.py [hunk 3]
```diff
@@ -979,7 +979,10 @@ def setUp(self):
     super().setUp()
     self.ys = np.array([[2.0, 0.1, -2.0], [0.3, -0.1, 1.2]], dtype=np.float32)
     self.ts = np.array([[0.0, 0.0, 1.0], [1.0, 0.0, 0.0]])
-    self._rtol = 5e-3 if jax.default_backend() != 'cpu' else 1e-6
+    # Relaxed tolerance to accommodate log-space numerical stability improvements
+    # The log-space focal loss implementation has slightly different numerical behavior
+    # for extreme values, which is expected and desirable for numerical stability
+    self._rtol = 5e-3 if jax.default_backend() != 'cpu' else 2e-5
```
Comment (vroulet) [inline optax/losses/_classification_test.py:985]: I would have expected the results to be more accurate. Is it because what is it tested against does not catch up the accuracy you gained?

File: optax/losses/_classification.py [hunk 4]
```diff
@@ -814,47 +815,88 @@ def sigmoid_focal_loss(
     alpha: Optional[float] = None,
     gamma: float = 2.0,
 ) -> chex.Array:
-  """Sigmoid focal loss.
+  """Sigmoid focal loss with numerical stability improvements.
```
Comment (vroulet) [inline optax/losses/_classification.py]: Add a r here so that you can simply use backslashes

File: optax/losses/_classification.py [hunk 5]
```diff
@@ -814,47 +815,88 @@ def sigmoid_focal_loss(
...
-  Use this loss function if classes are not mutually exclusive.
-  See `sigmoid_binary_cross_entropy` for more information.
+  The focal loss is a dynamically scaled cross entropy loss, where the scaling
+  factor decays to zero as confidence in the correct class increases. This
+  addresses class imbalance by down-weighting easy examples and focusing on
+  hard examples.
+
+  This implementation uses log-space computation for the focal weight
+  :math:`(1-p_t)^\\gamma` to ensure numerical stability, especially for
+  :math:`\\gamma < 2` and extreme logit values.
```
Comment (vroulet) [inline optax/losses/_classification.py]: just use \ instead of \\ in the docs (once you add a r in front of the ```)

**Review by vroulet — COMMENTED** [review]
That looks really great. I also appreciate a lot the additional documentation.
I'm leaving you a few comments, once resolved we should be able to merge.
Looking forward for more contributions :)

File: optax/losses/_classification_test.py (hunk 3 above)
Comment (leochlon) [reply optax/losses/_classification_test.py:985]: The failing test expected the focal loss to match a very specific mathematical formula for extreme logits with gamma=2. When we switched to the  log-space implementation implementation, we went from from using (1 - p_t) ** gamma directly to exp(gamma * log(1 - p_t)) , and our epsilon clamping prevents underflow, slightly changing results for extreme values
logsumexp precision. Finally, JAX's logsumexp has different floating-point behaviour than manual computation. For extreme logits, these have slightly different numerical behaviour log-space version is more numerically stable

File: optax/losses/_classification.py (hunk 2 above)
Comment (leochlon) [reply optax/losses/_classification.py]: So when we compute exp(γ * log(1 - p_t)), the gradient is:

∂/∂logits [exp(γ * log(1 - p_t))] = γ * (1 - p_t)^(γ-1) * ∂(1-p_t)/∂logits

When γ < 1: γ-1 < 0, so (1 - p_t)^(γ-1) = (1 - p_t)^{-(1-γ)} → ∞ as p_t → 1
When γ ≥ 1: γ-1 ≥ 0, so (1 - p_t)^(γ-1) → 0 as p_t → 1 (bounded)

Log-space computation solves numerical representation problems:

log(1 - p_t) is representable even when 1 - p_t underflows 
γ * log(1 - p_t) is representable even when (1 - p_t)^γ underflows

but this doesn't fix gradient explosion. Even if you can represent the values, the gradients can still explode when γ < 1, hessian explodes when γ < 2.

Clamping prevents gradient explosion by keeping p_t away from 1.

File: optax/losses/_classification.py (hunk 1 above)
Comment (leochlon) [reply optax/losses/_classification.py]: So I think the authors originally did this to use -1 as sentinel value to avoid Python control flow and make the compilation more JAX-friendly. But I get how it looks weird so I moved it to the end

File: optax/losses/_classification.py (hunk 5 above)
Comment (leochlon) [reply optax/losses/_classification.py]: done

File: optax/losses/_classification.py (hunk 2 above)
Comment (vroulet) [reply optax/losses/_classification.py]: It is pretty stable to me without the clamping (see code below).
Moving to logspace is sufficient for numerical stability. Of course from a mathematical viewpoint if log(1-pt) tends to -infty we would have issues. But numerically log(1-pt) (for e.g. label=0) is equivalent to -logits so unless we go to logits equal to huge numbers we are good. 

```python
import jax
import jax.numpy as jnp
import chex
import optax

def sigmoid_focal_loss(
    logits: chex.Array,
    labels: chex.Array,
    alpha = None,
    gamma: float = 2.0,
) -> chex.Array:
  chex.assert_type([logits], float)
  labels = jnp.astype(labels, logits.dtype)

  # Cross-entropy loss
  ce_loss = optax.sigmoid_binary_cross_entropy(logits, labels)

  # Compute log(1-p_t) using logsumexp unconditionally
  log_p = jax.nn.log_sigmoid(logits)
  log_q = jax.nn.log_sigmoid(-logits)

  log_one_minus_p_t = jax.scipy.special.logsumexp(
      jnp.stack([log_p, log_q], axis=-1),
      axis=-1,
      b=jnp.stack([1 - labels, labels], axis=-1)
  )

  # Focal weight and final loss
  focal_weight = jnp.exp(gamma * log_one_minus_p_t)
  loss = ce_loss * focal_weight

  # Alpha weighting
  if alpha is None:
      return loss
  else:
      weighted = (alpha * labels + (1 - alpha) * (1 - labels)) * loss
      return weighted

labels = jnp.asarray(0)
logits = jnp.asarray(-10.**4)
grad_stable = jax.grad(sigmoid_focal_loss)(logits, labels, gamma=0.5)
grad_unstable = jax.grad(optax.sigmoid_focal_loss)(logits, labels, gamma=0.5)
print(f'{grad_unstable=}, {grad_stable=} at {logits=}')
hess_stable = jax.grad(jax.grad(sigmoid_focal_loss))(logits, labels, gamma=0.5)
hess_unstable = jax.grad(jax.grad(optax.sigmoid_focal_loss))(logits, labels, gamma=0.5)
print(f'{hess_unstable=}, {hess_stable=} at {logits=}')
```

Comment (leochlon) [reply optax/losses/_classification.py]: You're right!!

The explosion in (1 - p_t)^(γ-1) is exactly canceled by the vanishing of ∂(1-p_t)/∂logits, resulting in a product that goes to 0!

For case $y=0$, $\text{logits} \to -\infty$ (the problematic regime):

The focal weight is $w = (\sigma(\text{logits}))^{\gamma}$ and its gradient is $$\frac{\partial w}{\partial \text{logits}} = \gamma \cdot (\sigma(\text{logits}))^{\gamma-1} \cdot \sigma(\text{logits})(1-\sigma(\text{logits}))$$
$$= \gamma \cdot (\sigma(\text{logits}))^{\gamma} \cdot (1-\sigma(\text{logits}))$$

While $(\sigma(\text{logits}))^{\gamma-1} \to \infty$ when $\gamma < 1$,  the full gradient contains the product $(\sigma(\text{logits}))^{\gamma-1} \cdot \sigma(\text{logits}) = (\sigma(\text{logits}))^{\gamma}$.

when we take the limit $\text{logits} \to -\infty$: $\sigma(\text{logits}) \sim e^{\text{logits}} \to 0$ and $(\sigma(\text{logits}))^{\gamma} \sim e^{\gamma \cdot \text{logits}} \to 0$ exponentially fast.

so $\lim_{\text{logits} \to -\infty} \frac{\partial w}{\partial \text{logits}} = 0$ for any $\gamma > 0$.

The explosion in $(\sigma(\text{logits}))^{\gamma-1}$ is exactly canceled by the vanishing of $\sigma(\text{logits})$, and so clamping is actually unnecessary in the logarithmic regime. 

new commit removes the clamping =]

File: optax/losses/_classification_test.py [hunk 6]
```diff
@@ -1069,6 +1072,44 @@ def test_ignore_negative(self):
     assert all(ce_loss[self.ts == 0] > 0)
     assert all(focal_loss[self.ts == 0] == 0)
 
+  @chex.all_variants
+  def test_extreme_logits_finite_gradients(self):
+    """Test that extreme logits with gamma < 1 produce finite gradients."""
+    # Test cases that previously caused NaN gradients
+    extreme_logits = jnp.array([25.0, -25.0, 50.0, -50.0])
```
Comment (vroulet) [inline optax/losses/_classification_test.py]: Go bigger, like 100. 
Have some non integer labels too.

File: optax/losses/_classification_test.py [hunk 7]
```diff
@@ -1069,6 +1072,44 @@ def test_ignore_negative(self):
...
+    labels = jnp.array([1.0, 0.0, 1.0, 0.0])
+
+    # Test gamma values in (0, 1) range
+    for gamma in [0.1, 0.5, 0.9]:
```
Comment (vroulet) [inline optax/losses/_classification_test.py]: Testing with a single gamma<1 (like gamma=0.5) is sufficient

File: optax/losses/_classification_test.py [hunk 8]
```diff
@@ -1069,6 +1072,44 @@ def test_ignore_negative(self):
...
+      def loss_fn(logits, test_gamma=gamma):
+        return jnp.sum(self.variant(_classification.sigmoid_focal_loss)(
+            logits, labels, gamma=test_gamma
+        ))
+
+      # Compute loss and gradients
+      loss_value = loss_fn(extreme_logits)
+      gradients = jax.grad(loss_fn)(extreme_logits)
```
Comment (vroulet) [inline optax/losses/_classification_test.py]: Could test with hessian too.

File: optax/losses/_classification_test.py [hunk 9]
```diff
@@ -1069,6 +1072,44 @@ def test_ignore_negative(self):
...
+
+      # Verify that both loss and gradients are finite
+      self.assertTrue(jnp.isfinite(loss_value),
+                      f"Loss should be finite for gamma={gamma}, got {loss_value}")
+      self.assertTrue(jnp.all(jnp.isfinite(gradients)),
+                      f"Gradients should be finite for gamma={gamma}, got {gradients}")
+
+    # Also test with alpha weighting
+    def alpha_loss_fn(logits):
```
Comment (vroulet) [inline optax/losses/_classification_test.py]: You can remove the test with alpha. It won't pose problems

**Review by vroulet — COMMENTED** [review]
Great. Some additional comments on the tests.
Also check that it matches the sigmoid_binary_cross_entropy() for gamma =0

File: optax/losses/_classification_test.py (hunk 6 above)
Comment (leochlon) [reply optax/losses/_classification_test.py]: Done

File: optax/losses/_classification_test.py (hunk 7 above)
Comment (leochlon) [reply optax/losses/_classification_test.py]: Done

File: optax/losses/_classification_test.py (hunk 8 above)
Comment (leochlon) [reply optax/losses/_classification_test.py]: There was a Hessian test in the same function, did you mean a different kind of test?

  # Test Hessians for numerical stability
  hessian = jax.hessian(loss_fn)(extreme_logits)
  self.assertTrue(jnp.all(jnp.isfinite(hessian)),
                  f"Hessians should be finite for gamma={gamma}")

File: optax/losses/_classification_test.py (hunk 9 above)
Comment (leochlon) [reply optax/losses/_classification_test.py]: done

**Comment by rdyro** [comment]
Could you squash the commits please, when you're ready with the PR?

**Comment by leochlon** [comment]
@rdyro  done.

**Comment by leochlon** [comment]
@Vishal-sys-code sorry accident. could you please tell me how to merge now that everyone approved it?

**Comment by vroulet** [comment]
@leochlon, your PR will be be merged internally.
Maintainers of a library are the ones responsible for approving and merging PRs.

@Vishal-sys-code, could you explain what is happening? Is it a class?
It would have been best from the start to contact us and explain us what you are doing.

**Comment by Vishal-sys-code** [comment]
@vroulet 

**What's happening**: This PR fixes numerical instability in `sigmoid_focal_loss` where extreme logits (like ±100) with gamma < 2 caused NaN gradients. @leochlon rewrote it using log-space computation for stability.

**Is it a class?**: No, this is a function fix, `sigmoid_focal_loss` is a standalone function in the losses module, not a class.

**Key changes:**
- 92 lines changed in _classification.py (66 additions, 26 deletions)
- 56 lines changed in tests (55 additions, 1 deletion)
- Maintains backward compatibility while fixing numerical issues
- Adds support for continuous labels in [0,1] range

This appears to be a legitimate and well-implemented fix for a real numerical stability issue.

**Comment by vroulet** [comment]
There seems to be a misunderstanding. 
Why are you reviewing, approving PRs on this library? 
@leochlon, why are you asking reviews from @Vishal-sys-code and why are you asking him how to merge?

**Comment by leochlon** [comment]
@vroulet I’m confused, I thought they worked at DeepMind with you guys.

**Comment by rdyro** [comment]
> @vroulet I’m confused, I thought they worked at DeepMind with you guys.

No, @Vishal-sys-code is not affiliated with us, I'm not sure why they provided their review.

**Comment by leochlon** [comment]
@rdyro Ah I see why everyone was confused now. 

Btw @vroulet the effort required to refactor the other two PRs for Optimistix just isn't worth the extra effort, so I'll leave the code here and one day if you guys ever do change your mind about the scope, feel free to reopen the PRs again. If you ever do end up using it just put me somewhere in the credits I don't mind.
